### PR TITLE
Re:Cargo fixes

### PR DESCRIPTION
**Pull request by AlexMorgan3817**
## Changelog
:cl:
tweak: McRonalds now buy not roaches, but their meat.
fix: fixed that special offers wasn't able to be fulfilled in some situations
fix: special offers won't able to request 0 from 0 things
fix: fixed freezing of UI after first buy
add: You can now sell things to trade stations
add: Trade beacons now on activation not just animated, but have sparks. If you will be on tile where goods will "land", you will be injured.
add: Now you can't farm crates for free, if you buy goods with crates you will pay for crate as commision
add: Added cat ears that can be bought in McRonalds together with Chocola cake
add: You can now input how many goods you want to add in cart.
spellcheck: Fixed typos on GBT 'Suit Up'
spellcheck: FSMS 'Kaida' fixed unfinished descri
/:cl:

### CODE REVIEW

**Comment by AlexMorgan3817** [comment]
Put here mapping label

**Comment by AlexMorgan3817** [comment]
Понял

**Comment by AlexMorgan3817** [comment]
Wait, when i convert this from draft...

File: code/modules/economy/price_list.dm [hunk 1]
```diff
@@ -1031,15 +1031,6 @@
...
 	price_tag = 500
 
-/mob/living/carbon/superior_animal/roach
-	price_tag = 150
-
-/mob/living/carbon/superior_animal/roach/roachling
-	price_tag = 100
-
-/mob/living/carbon/superior_animal/roach/hunter
-	price_tag = 200
```
Comment (Hopekz) [inline code/modules/economy/price_list.dm:1041]: Are unprocessed roaches going to be factored in now that you removed this part?

File: code/modules/cargo/exports/sheets.dm [hunk 2]
```diff
@@ -145,7 +145,7 @@
 // Uranium. Still useful for both power generation and nuclear annihilation.
 /datum/export/stack/uranium
 	cost = 20
-	export_types = list(/obj/item/stack/material/diamond)
+	export_types = list(/obj/item/stack/material/uranium)
```
Comment (Hopekz) [inline code/modules/cargo/exports/sheets.dm:148]: Good oversight fix

File: code/modules/economy/price_list.dm (hunk 1 above)
Comment (Hopekz) [inline code/modules/economy/price_list.dm:1041]: Are unprocessed roaches going to be factored in now that you removed this part?

**Review by Hopekz — APPROVED** [review]
nhp8v4DVda_nice-export

File: code/modules/economy/price_list.dm (hunk 1 above)
Comment (AlexMorgan3817) [reply code/modules/economy/price_list.dm:1041]: What?

**Comment by AlexMorgan3817** [comment]
Done

**Comment by AlexMorgan3817** [comment]
Ready for review

**Comment by TorinoFermic** [comment]
I would like to have the changelog in your first post

File: code/modules/modular_computers/file_system/programs/generic/trade.dm [hunk 3]
```diff
@@ -150,7 +162,7 @@
 		ui.set_initial_data(data)
 		ui.open()
 
-/datum/nano_module/program/trade/ui_data()
+/datum/nano_module/program/trade/ui_data()
```
Comment (AlexMorgan3817) [inline code/modules/modular_computers/file_system/programs/generic/trade.dm:165]:
```suggestion
/datum/nano_module/program/trade/ui_data()
```

**Comment by AlexMorgan3817** [comment]
Changelog done

**Comment by AlexMorgan3817** [comment]
Conficts fixed

**Comment by AlexMorgan3817** [comment]
Dudes, gde review or merge?

File: code/controllers/subsystems/trade.dm [hunk 4]
```diff
@@ -1,5 +1,5 @@
 SUBSYSTEM_DEF(trade)
-	name = "Trade"
+	name = "Asters Automated Trading System"
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:2]: Please don't, the name's for a debug system, and is not visible to the players.

File: code/controllers/subsystems/trade.dm [hunk 5]
```diff
@@ -174,38 +176,69 @@ SUBSYSTEM_DEF(trade)
...
+				if(isnum(tcount))
+					. += tcount
+
+/datum/controller/subsystem/trade/proc/collect_price_for_list(list/m)
+	. = 0
+	for(var/t in m)
+		if(ispath(t))
+			. += get_import_cost(t) * get_2d_matrix_cell(m, t, "count")
+
+/datum/controller/subsystem/trade/proc/shoplist2list(list/m)
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:195]: Please, for the sanity of people in the future who will have to debug this, give your variables better names than m, ., dot, i, il, and b

File: code/controllers/subsystems/trade.dm [hunk 6]
```diff
@@ -174,38 +176,69 @@ SUBSYSTEM_DEF(trade)
...
+	if(islist(m))
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:196]: I would sincerely hope m is a list, considering you have a list check already in the buy proc.

File: code/controllers/subsystems/trade.dm [hunk 7]
```diff
@@ -174,38 +176,69 @@ SUBSYSTEM_DEF(trade)
...
+		. = list()
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:197]: I know you're being fancy with pointers, but please, just return dot (or hopefully whatever you rename it to).

File: code/controllers/subsystems/trade.dm [hunk 8]
```diff
@@ -174,38 +176,69 @@ SUBSYSTEM_DEF(trade)
...
-	for(var/category_name in shoppinglist)
-		var/list/category = shoppinglist[category_name]
-		for(var/path in category)
-			cost += get_import_cost(path, station) * category[path]
-
+	var/obj/structure/closet/crate/C
+	var/list/sl = shoplist2list(shoppinglist)
+	var/count_of_all = collect_counts_from(sl)
+	cost = collect_price_for_list(sl)
+	if(count_of_all > 1 && count_of_all)
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:218]: count_of_all wouldn't be null if count_of_all is greater than 1

File: code/controllers/subsystems/trade.dm [hunk 9]
```diff
@@ -174,38 +176,69 @@ SUBSYSTEM_DEF(trade)
...
+	if(QDELETED(beacon) || !account || !recursiveLen(shoppinglist) || !istype(station))
 		return
 
-	var/cost = 0
-	for(var/category_name in shoppinglist)
-		var/list/category = shoppinglist[category_name]
-		for(var/path in category)
-			cost += get_import_cost(path, station) * category[path]
-
+	var/obj/structure/closet/crate/C
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:214]: Again, give your variables better names so it's easier to debug and review.

File: code/game/objects/items/devices/biosyphon.dm [hunk 10]
```diff
@@ -17,8 +17,8 @@
 	var/last_produce = 0
 	var/cooldown = 15 MINUTES
 
-/obj/item/biosyphon/New()
-	..()
+/obj/item/biosyphon/Initialize()
```
Comment (MadmanMartian) [inline code/game/objects/items/devices/biosyphon.dm:20]: Unatomic

File: code/modules/cargo/exports/sheets.dm (hunk 2 above)
Comment (MadmanMartian) [inline code/modules/cargo/exports/sheets.dm:148]: Unatomic

File: code/modules/trade/datums/_trade_station.dm [hunk 11]
```diff
@@ -15,6 +15,7 @@
 	var/spawn_cost = 1
 	var/start_discovered = FALSE
 	var/list/linked_with //trade 'stations' or 'station' that must spawn with //list or path
+	var/commision = 200
```
Comment (MadmanMartian) [inline code/modules/trade/datums/_trade_station.dm:18]: Add a helpful comment detailing what this variable does.

File: code/modules/trade/datums/trade_stations_presets/McRonalds.dm [hunk 12]
```diff
@@ -9,7 +9,8 @@
 		"Misc" = list(
 			/obj/item/weapon/reagent_containers/food/snacks/fishandchips = custom_good_name("Duchess Fish 'n' Lord Chips"),
 			/obj/item/weapon/reagent_containers/food/snacks/sliceable/plaincake = custom_good_name("Vanilla"),
-			/obj/item/weapon/reagent_containers/food/snacks/sliceable/chocolatecake = custom_good_name("Chocola")
+			/obj/item/weapon/reagent_containers/food/snacks/sliceable/chocolatecake = custom_good_name("Chocola"),
+			/obj/item/clothing/head/kitty = custom_good_name("Definitely not Chocola")
```
Comment (MadmanMartian) [inline code/modules/trade/datums/trade_stations_presets/McRonalds.dm:13]: image

**Comment by AlexMorgan3817** [comment]
Done ._.

**Review by TorinoFermic — APPROVED** [review]

File: code/modules/trade/datums/trade_stations_presets/McRonalds.dm (hunk 12 above)
Comment (MadmanMartian) [inline code/modules/trade/datums/trade_stations_presets/McRonalds.dm:13]: No.

**Review by MadmanMartian — CHANGES_REQUESTED** [review]
No.

File: code/modules/trade/datums/trade_stations_presets/McRonalds.dm (hunk 12 above)
Comment (genessee596) [inline code/modules/trade/datums/trade_stations_presets/McRonalds.dm:13]: No.

**Review by genessee596 — CHANGES_REQUESTED** [review]

File: code/modules/trade/datums/trade_stations_presets/McRonalds.dm (hunk 12 above)
Comment (Firefox13) [inline code/modules/trade/datums/trade_stations_presets/McRonalds.dm:13]: No.

File: code/modules/trade/datums/trade_stations_presets/McRonalds.dm [hunk 13]
```diff
@@ -21,7 +22,8 @@
 	offer_types = list(
 		/obj/item/weapon/reagent_containers/food/snacks/meat,
 		/obj/item/weapon/reagent_containers/food/snacks/meat/corgi,
-		/mob/living/carbon/superior_animal/roach,
-		/mob/living/carbon/superior_animal/roach/hunter,
-		/mob/living/carbon/superior_animal/roach/roachling
+		/obj/item/weapon/reagent_containers/food/snacks/meat/roachmeat,
+		/obj/item/weapon/reagent_containers/food/snacks/meat/roachmeat/seuche,
+		/obj/item/weapon/reagent_containers/food/snacks/meat/roachmeat/kraftwerk,
+		/obj/item/weapon/reagent_containers/food/snacks/meat/roachmeat/jager,
```
Comment (Firefox13) [inline code/modules/trade/datums/trade_stations_presets/McRonalds.dm:28]: Запятая перед скобкой

File: code/modules/trade/datums/trade_stations_presets/McRonalds.dm (hunk 12 above)
Comment (Wouju) [inline code/modules/trade/datums/trade_stations_presets/McRonalds.dm:13]: Yes.

I'm joking, **No**

**Comment by TheShown911** [comment]
I approve the ears as they are already found in the neko, but there is too much opposition to merge yet

**Comment by genessee596** [comment]
Please change the neko catear thing. I won't be approving this PR until that gets fixed.

File: code/game/objects/items/weapons/storage/boxes.dm [hunk 14]
```diff
@@ -595,3 +595,23 @@
...
+	icon_state = "happy_meal"
+
+/obj/item/weapon/storage/box/happy_meal/New()
+	..()
+	var/list/things2spawn = list(
+		/obj/item/weapon/reagent_containers/food/snacks/sliceable/plaincake,
+		/obj/item/weapon/reagent_containers/food/snacks/sliceable/chocolatecake,
+		/obj/item/weapon/reagent_containers/food/snacks/bigbiteburger,
+		/obj/item/weapon/reagent_containers/food/snacks/fishandchips,
+		/obj/item/clothing/head/kitty
```
Comment (MadmanMartian) [inline code/game/objects/items/weapons/storage/boxes.dm:611]: No.

**Review by MadmanMartian — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/storage/boxes.dm [hunk 15]
```diff
@@ -595,3 +595,23 @@
...
+		)
```
Comment (AlexMorgan3817) [inline code/game/objects/items/weapons/storage/boxes.dm:612]:
```suggestion
	var/list/things2spawn = list(
		/obj/item/weapon/reagent_containers/food/snacks/sliceable/plaincake,
		/obj/item/weapon/reagent_containers/food/snacks/sliceable/chocolatecake,
		/obj/item/weapon/reagent_containers/food/snacks/bigbiteburger,
		/obj/item/weapon/reagent_containers/food/snacks/fishandchips
		)
		if(prob(1))
			things2spawn += /obj/item/clothing/head/kitty
```

File: code/game/objects/items/weapons/storage/boxes.dm [hunk 16]
```diff
@@ -595,3 +595,23 @@
...
+/obj/item/weapon/storage/box/happy_meal/New()
+	..()
+	var/list/things2spawn = list(
+		/obj/item/weapon/reagent_containers/food/snacks/sliceable/plaincake,
+		/obj/item/weapon/reagent_containers/food/snacks/sliceable/chocolatecake,
+		/obj/item/weapon/reagent_containers/food/snacks/bigbiteburger,
+		/obj/item/weapon/reagent_containers/food/snacks/fishandchips
+		)
+		if(prob(1))
+			things2spawn += /obj/item/clothing/head/kitty
```
Comment (AlexMorgan3817) [inline code/game/objects/items/weapons/storage/boxes.dm:613]:
```suggestion
	var/list/things2spawn = list(
		/obj/item/weapon/reagent_containers/food/snacks/sliceable/plaincake,
		/obj/item/weapon/reagent_containers/food/snacks/sliceable/chocolatecake,
		/obj/item/weapon/reagent_containers/food/snacks/bigbiteburger,
		/obj/item/weapon/reagent_containers/food/snacks/fishandchips
	)
	if(prob(1))
		things2spawn += /obj/item/clothing/head/kitty
```

File: code/game/turfs/turf.dm [hunk 17]
```diff
@@ -236,16 +236,14 @@ var/const/enterloopsanity = 100
 				L.Add(t)
 	return L
 
-/turf/proc/contains_dense_objects()
+/turf/proc/contains_dense_objects(unincludehumans)
 	if(density)
 		return 1
 	for(var/atom/A in src)
-		if(A.density && !(A.flags & ON_BORDER))
+		if(A.density && !(A.flags & ON_BORDER) && (unincludehumans && !ishuman(A)))
```
Comment (AlexMorgan3817) [inline code/game/turfs/turf.dm:243]:
```suggestion
		if(A.density && !(A.flags & ON_BORDER) || (unincludehumans && !ishuman(A)))
```

**Comment by TheShown911** [comment]
Conflicts

**Review by genessee596 — APPROVED** [review]
i see no cat ears'

neuron activated

File: code/controllers/subsystems/trade.dm [hunk 18]
```diff
@@ -38,14 +39,14 @@ SUBSYSTEM_DEF(trade)
 	var/list/stations2init = collect_spawn_always()
 
 	while(trade_stations_budget && length(weightstationlist))
-		var/datum/trade_station/s = pickweight(weightstationlist)
-		if(istype(s))
-			stations2init += s
-		s.cost_trade_stations_budget()
-		weightstationlist.Remove(s)
+		var/datum/trade_station/station_instanse = pickweight(weightstationlist)
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:42]: It's instance, not instanse, let's not have a _doble situation.

File: code/controllers/subsystems/trade.dm [hunk 19]
```diff
@@ -125,7 +126,7 @@ SUBSYSTEM_DEF(trade)
 		if(AM.anchored)
 			continue
 
-		var/export_cost = get_export_cost(AM)
+		var/export_cost = get_export_cost(AM) * 1.25
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:129]: Explain this arbitrary number.

File: code/controllers/subsystems/trade.dm [hunk 20]
```diff
@@ -174,38 +175,72 @@ SUBSYSTEM_DEF(trade)
 
 	station.generate_offer()
 
+/datum/controller/subsystem/trade/proc/collect_counts_from(list/m)
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:178]: Comment what this function does, as your variable names are nondescript.

File: code/game/objects/items/weapons/storage/boxes.dm [hunk 21]
```diff
@@ -595,3 +595,22 @@
...
+/obj/item/weapon/storage/box/happy_meal/New()
+	..()
+	var/list/things2spawn = list(
+		/obj/item/weapon/reagent_containers/food/snacks/sliceable/plaincake,
+		/obj/item/weapon/reagent_containers/food/snacks/sliceable/chocolatecake,
+		/obj/item/weapon/reagent_containers/food/snacks/bigbiteburger,
+		/obj/item/weapon/reagent_containers/food/snacks/fishandchips
+	)
+	if(prob(1))
+		things2spawn += /obj/item/clothing/head/kitty
```
Comment (MadmanMartian) [inline code/game/objects/items/weapons/storage/boxes.dm:613]: yCpgFHz

**Review by MadmanMartian — CHANGES_REQUESTED** [review]

**Comment by MadmanMartian** [comment]
And stop marking stuff as resolved when you haven't even addressed them.

**Comment by AlexMorgan3817** [comment]
Finally, only conflicts left... and maybe cat ears removing... 
image

File: code/game/objects/items/weapons/storage/boxes.dm (hunk 21 above)
Comment (AlexMorgan3817) [reply code/game/objects/items/weapons/storage/boxes.dm:668]: done

File: code/controllers/subsystems/trade.dm (hunk 20 above)
Comment (AlexMorgan3817) [reply code/controllers/subsystems/trade.dm]: Done

**Comment by AlexMorgan3817** [comment]
Done

**Comment by AlexMorgan3817** [comment]
Ready for merge and reviews

File: code/controllers/subsystems/trade.dm [hunk 22]
```diff
@@ -38,14 +39,14 @@ SUBSYSTEM_DEF(trade)
...
+		var/datum/trade_station/station_instance = pickweight(weightstationlist)
+		if(istype(station_instance))
+			stations2init += station_instance
+			station_instance.cost_trade_stations_budget()
+		weightstationlist.Remove(station_instance)
 	init_stations_by_list(stations2init)
 
-/datum/controller/subsystem/trade/proc
+/datum/controller/subsystem/trade/proc/
 	collect_trade_stations()
```
Comment (MadmanMartian) [inline code/controllers/subsystems/trade.dm:50]: This is incredibly heretical

Comment (AlexMorgan3817) [reply code/controllers/subsystems/trade.dm:50]: image

File: code/modules/trade/datums/trade_stations_presets/fs_weapon_factory.dm [hunk 23]
```diff
@@ -1,5 +1,5 @@
 /datum/trade_station/fs_factory
-	name_pool = list("FSMS 'Kaida'" = "Frozen Star Manufacturing Station 'Kaida'. It ")
+	name_pool = list("FSMS 'Kaida'" = "Frozen Star Manufacturing Station 'Kaida'. Maybe they have an extra batch of weapon?")
```
Comment (MadmanMartian) [inline code/modules/trade/datums/trade_stations_presets/fs_weapon_factory.dm:2]: "Maybe they have an extra batch of weapons?"

**Review by MadmanMartian — APPROVED** [review]
Besides that tag and icon state re-order that's happened in the mapping, all looks well.

**Review by Evankhell561 — CHANGES_REQUESTED** [review]
it looks like you didn't use map marge.

**Comment by AlexMorgan3817** [comment]
> it looks like you didn't use map marge.

Suka, blyat

**Comment by AlexMorgan3817** [comment]
> it looks like you didn't use map marge.

@eva, no, it is used

**Comment by Evankhell561** [comment]
> > it looks like you didn't use map marge.
> 
> @Evankhell561 , no, it is used

image

**Comment by MadmanMartian** [comment]
Check their eris file, it just looks like the mapcleaner messed up and changed the order of tag

**Comment by Kurgis** [comment]
It appears as though it was rearranged into alphabetical order, though I'm unsure if that's just coincidence.

**Comment by Kurgis** [comment]
Looking through the file again, something sparked through my head seeing the `tag` var used. I looked around to confirm my suspicions, and found this quote from https://tgstation13.org/wiki/Guide_to_mapping:

> While building your station you might find yourself in need of an object with a specific alignment (windows, pipes, cables etc) that does not appear in your menu. This is because BYOND does not simulate instances of an object other than its base state unless they are present on the map. You can generate them by right clicking an object and selecting "generate instance from state/direction". **This function has an unwanted feature in which a generated object sometimes has a tag added automatically. These tags can cause error with certain features of the game during a round, and need to be removed.**

While it's certainly not an issue for this PR, we might want to see to removing these tag vars at some point later in the future.

**Comment by AlexMorgan3817** [comment]
> > > it looks like you didn't use map marge.
> > 
> > 
> > @Evankhell561 , no, it is used
> 
> image

image

**Comment by AlexMorgan3817** [comment]
> > > it looks like you didn't use map marge.
> > 
> > 
> > @Evankhell561 , no, it is used
> 
> image

image
image

**Comment by TorinoFermic** [comment]
Everything check out, not a bad map problem here.

**Pull request merged**